# Data Cleaning 

In [159]:
# Read in the customer data file using Python in Jupyter Notebook
import numpy as np
import pandas as pd
# Start by importing numpy and pandas

In [606]:
df1 = pd.read_csv(input(''))
df1
# customer_list_updated.csv

 customer_list_updated.csv


cust_id|date|time|name|email|phone|sms-opt-out 
0    1|2023-03-15|08:45:12|Rachel|rachel@centralper...
1    2|2023-05-22|12:30:45|R&! Geller|rossg@central...
2    3|2023-07-09|18:15:27|Monica Geller|chefmonica...
3    4|2023-09-01|21:05:33|Chandler Bing|chandlerb@...
4    5|2023-11-18|14:22:10|Joey|howyoudoing@central...
..                                                 ...
516  517|2024-02-15|11:30:30|Perry the Platypus|age...
517  518|2024-05-01|15:55:45|Baljeet|baljeet@danvil...
518  519|2023-03-05|09:20:00|Buford|buford@danville...
519  520|2023-05-18|13:45:15|Vanessa D.^|vanessa@da...
520  521|2023-08-01|18:10:30|Linda Flynn-Fletcher|l...

[521 rows x 1 columns]

*To load the data as a DataFrame will use the above code*
*Asking the user to input the file name for the file to be processed*
*Assign to a variable to save it into a dataframe*
*The data is in an awkward format will need to separate customer records into columns*

In [608]:
df1 = pd.read_csv(input(''), sep = '|')
df1
#customer_list_updated.csv

 customer_list_updated.csv


cust_id        date      time                  name  \
0          1  2023-03-15  08:45:12                Rachel   
1          2  2023-05-22  12:30:45            R&! Geller   
2          3  2023-07-09  18:15:27         Monica Geller   
3          4  2023-09-01  21:05:33         Chandler Bing   
4          5  2023-11-18  14:22:10                  Joey   
..       ...         ...       ...                   ...   
516      517  2024-02-15  11:30:30    Perry the Platypus   
517      518  2024-05-01  15:55:45               Baljeet   
518      519  2023-03-05  09:20:00                Buford   
519      520  2023-05-18  13:45:15           Vanessa D.^   
520      521  2023-08-01  18:10:30  Linda Flynn-Fletcher   

                              email         phone sms-opt-out   
0         rachel@centralperk.coffee  212-555-1001            N  
1          rossg@centralperk.coffee  212-555-1002            N  
2     chefmonica@centralperk.coffee  212-555-1003            N  
3      chandlerb@centralperk.coffee  212-555-1004            Y  
4    howyoudoing@centralperk.coffee  212-555-1005            N  
..                              ...           ...          ...  
516         agentp@danvillekids.com  555-555-3206            N  
517        baljeet@danvillekids.com  555-555-3207            N  
518         buford@danvillekids.com  555-555-3208            N  
519        vanessa@danvillekids.com  555-555-3209            N  
520          linda@danvillekids.com  555-555-3210            N  

[521 rows x 7 columns]

*In order to separate the customer recordsinto columns I used the separator parameter to display the values without the pipe symbol*
*Each record is then separated where the pipe symbol is located and separated into their perspective columns*
*Now, if I do it this way, I can make it easier to preview the data*
*call on the variable to display*

In [610]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 521 entries, 0 to 520
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   cust_id       521 non-null    int64 
 1   date          521 non-null    object
 2   time          521 non-null    object
 3   name          521 non-null    object
 4   email         521 non-null    object
 5   phone         520 non-null    object
 6   sms-opt-out   520 non-null    object
dtypes: int64(1), object(6)
memory usage: 28.6+ KB


*Checking the structure after having separated the customer records into columns using the info attribute*
*This will show the summary infromation about df1 to help me identify any irregularities in the columns*
*Here I can tell I have 521 records*
*I can tell have a total of 7 columns*
*The Dtype for most of this data is string except for the customer id which an integer of 64 bytes*
*The amount of non-null values or any value that is not none is 521 entries compared to 520 non null values in the message opt out column where there could be a missing value and 520 phone non null values in phone,
which means there could be a missing phone number int hat column*
*I will then need to check the contents*

In [611]:
df1.head(51)

cust_id        date      time            name  \
0         1  2023-03-15  08:45:12          Rachel   
1         2  2023-05-22  12:30:45      R&! Geller   
2         3  2023-07-09  18:15:27   Monica Geller   
3         4  2023-09-01  21:05:33   Chandler Bing   
4         5  2023-11-18  14:22:10            Joey   
5         6  2024-01-05  10:55:49      P&! Buffay   
6         7  2024-02-14  16:40:05         Gunther   
7         8  2023-04-20  09:15:30          Janice   
8         9  2023-06-30  13:50:55        Mike H.^   
9        10  2023-08-25  17:25:10           Emily   
10       11  2023-04-12  09:20:30    Walter White   
11       12  2023-06-25  13:45:50   Jesse Pinkman   
12       13  2023-08-03  17:30:15      Skyler W.^   
13       14  2023-10-19  20:05:40    Saul Goodman   
14       15  2024-02-28  11:15:25    H&! Schrader   
15       16  2024-03-10  15:50:35       Gus Fring   
16       17  2024-04-22  19:25:50        Mike E.^   
17       18  2023-05-14  10:30:45       Marie S.^   
18       19  2023-07-27  14:05:00            Todd   
19       20  2023-09-09  18:40:15       Lydia R.^   
20       21  2023-01-05  06:15:20        Jon Snow   
21       22  2023-02-18  10:40:35    Daenerys T.^   
22       23  2023-04-22  14:05:50      Tyrion L.^   
23       24  2023-06-30  18:30:10      Arya Stark   
24       25  2023-08-15  22:55:25       Sansa S.^   
25       26  2023-10-28  07:20:40      Cersei L.^   
26       27  2023-12-12  11:45:55   J&! Lannister   
27       28  2023-03-03  08:10:15      Bran Stark   
28       29  2023-05-17  12:35:30     Samwell T.^   
29       30  2023-07-29  16:00:45         Brienne   
30       31  2023-03-10  09:25:30       Homer S.^   
31       32  2023-05-05  13:50:45           Marge   
32       33  2023-07-20  18:15:00            Bart   
33       34  2023-09-15  22:40:15        Lisa S.^   
34       35  2023-11-08  07:05:30          Maggie   
35       36  2024-01-25  11:30:45    Ned Flanders   
36       37  2024-03-18  15:55:00      Mr&! Burns   
37       38  2023-04-22  10:20:15          Krusty   
38       39  2023-06-14  14:45:30    Chief Wiggum   
39       40  2023-08-27  19:10:45        Milhouse   
40       41  2023-02-12  08:35:15   Michael Scott   
41       42  2023-04-28  12:00:30         Jim H.^   
42       43  2023-07-05  16:25:45      Pam Beesly   
43       44  2023-09-22  20:50:00  Dwight Schrute   
44       45  2023-11-30  06:15:15      Angela M.^   
45       46  2024-02-14  10:40:30       Kevin M.^   
46       47  2024-04-01  15:05:45     Stanley H.^   
47       48  2023-03-18  09:30:00        Andy B.^   
48       49  2023-05-25  13:55:15       Oscar M.^   
49       50  2023-08-10  18:20:30           Creed   
50       51  2023-03-22  07:15:20          Eleven   

                                   email         phone sms-opt-out   
0              rachel@centralperk.coffee  212-555-1001            N  
1               rossg@centralperk.coffee  212-555-1002            N  
2          chefmonica@centralperk.coffee  212-555-1003            N  
3           chandlerb@centralperk.coffee  212-555-1004            Y  
4         howyoudoing@centralperk.coffee  212-555-1005            N  
5           smellycat@centralperk.coffee  212-555-1006            N  
6             gunther@centralperk.coffee  212-555-1007            N  
7             ohmygod@centralperk.coffee  212-555-1008            N  
8                mike@centralperk.coffee  212-555-1009            N  
9               emily@centralperk.coffee  212-555-1010            N  
10      heisenberg@lospolloshermanos.biz  505-555-2001            N  
11           jesse@lospolloshermanos.biz  505-555-2002            N  
12          skyler@lospolloshermanos.biz  505-555-2003            Y  
13  bettercallsaul@lospolloshermanos.biz  505-555-2004            N  
14            hank@lospolloshermanos.biz  505-555-2005            N  
15             gus@lospolloshermanos.biz  505-555-2006            N  
16            mike@lospolloshermanos.biz  505-555-2007   

*looking at the first 50 records with .head() attribute. With this, I'm able to see that the name values have stray characters*
*Will need to take out the caret symbol(^) also the (&!) symbols*
*Format the phone numbers correctly*
*Decide what to do with the missing valyes in the message opt out column*

In [614]:
df1["name"] = df1["name"].str.replace("^","")
df1

cust_id        date      time                  name  \
0          1  2023-03-15  08:45:12                Rachel   
1          2  2023-05-22  12:30:45             R. Geller   
2          3  2023-07-09  18:15:27         Monica Geller   
3          4  2023-09-01  21:05:33         Chandler Bing   
4          5  2023-11-18  14:22:10                  Joey   
..       ...         ...       ...                   ...   
516      517  2024-02-15  11:30:30    Perry the Platypus   
517      518  2024-05-01  15:55:45               Baljeet   
518      519  2023-03-05  09:20:00                Buford   
519      520  2023-05-18  13:45:15            Vanessa D.   
520      521  2023-08-01  18:10:30  Linda Flynn-Fletcher   

                              email         phone sms-opt-out   
0         rachel@centralperk.coffee  212-555-1001            N  
1          rossg@centralperk.coffee  212-555-1002            N  
2     chefmonica@centralperk.coffee  212-555-1003            N  
3      chandlerb@centralperk.coffee  212-555-1004            Y  
4    howyoudoing@centralperk.coffee  212-555-1005            N  
..                              ...           ...          ...  
516         agentp@danvillekids.com  555-555-3206            N  
517        baljeet@danvillekids.com  555-555-3207            N  
518         buford@danvillekids.com  555-555-3208            N  
519        vanessa@danvillekids.com  555-555-3209            N  
520          linda@danvillekids.com  555-555-3210            N  

[521 rows x 7 columns]

*I replaced the caret symbol (^) first by pulling the name column in brackets, this will specify the column I want to make changes to*
*I then applied the .str method to make sure the output we get in return is also a string, along with the. replace method to replace the caret symbol with nothing since its at the very end of the name string.*
*I then went ahead and assigned it to the same dataframe and same column to save the changes*
*Finally I called on it to view the output*

In [616]:
df1["name"] = df1["name"].str.replace("&!",".")
df1

cust_id        date      time                  name  \
0          1  2023-03-15  08:45:12                Rachel   
1          2  2023-05-22  12:30:45             R. Geller   
2          3  2023-07-09  18:15:27         Monica Geller   
3          4  2023-09-01  21:05:33         Chandler Bing   
4          5  2023-11-18  14:22:10                  Joey   
..       ...         ...       ...                   ...   
516      517  2024-02-15  11:30:30    Perry the Platypus   
517      518  2024-05-01  15:55:45               Baljeet   
518      519  2023-03-05  09:20:00                Buford   
519      520  2023-05-18  13:45:15            Vanessa D.   
520      521  2023-08-01  18:10:30  Linda Flynn-Fletcher   

                              email         phone sms-opt-out   
0         rachel@centralperk.coffee  212-555-1001            N  
1          rossg@centralperk.coffee  212-555-1002            N  
2     chefmonica@centralperk.coffee  212-555-1003            N  
3      chandlerb@centralperk.coffee  212-555-1004            Y  
4    howyoudoing@centralperk.coffee  212-555-1005            N  
..                              ...           ...          ...  
516         agentp@danvillekids.com  555-555-3206            N  
517        baljeet@danvillekids.com  555-555-3207            N  
518         buford@danvillekids.com  555-555-3208            N  
519        vanessa@danvillekids.com  555-555-3209            N  
520          linda@danvillekids.com  555-555-3210            N  

[521 rows x 7 columns]

*I then replaced the (&) and (!) symbol together since I seen that both of them are together for the first 50 records.* 
*This time I replaced it with a period since I noticed they were in between a first intial followed by a last name*
*I again pulled up the name column in brackets, this will specify the column I want to make changes to*
*I then applied the .str method again to make sure the output we get in return is also a string, along with the. replace method*
*I then went ahead and assigned it to the same dataframe and same column to save the changes*
*Finally, I called on it to view the output*

In [587]:
#df1_cleaned = df1

In [588]:
#df1_cleaned["phone"] = df1["phone"].astype(str)
#df1_cleaned

cust_id        date      time                  name  \
0          1  2023-03-15  08:45:12                Rachel   
1          2  2023-05-22  12:30:45             R. Geller   
2          3  2023-07-09  18:15:27         Monica Geller   
3          4  2023-09-01  21:05:33         Chandler Bing   
4          5  2023-11-18  14:22:10                  Joey   
..       ...         ...       ...                   ...   
516      517  2024-02-15  11:30:30    Perry the Platypus   
517      518  2024-05-01  15:55:45               Baljeet   
518      519  2023-03-05  09:20:00                Buford   
519      520  2023-05-18  13:45:15            Vanessa D.   
520      521  2023-08-01  18:10:30  Linda Flynn-Fletcher   

                              email         phone sms-opt-out   
0         rachel@centralperk.coffee  212-555-1001            N  
1          rossg@centralperk.coffee  212-555-1002            N  
2     chefmonica@centralperk.coffee  212-555-1003            N  
3      chandlerb@centralperk.coffee  212-555-1004            Y  
4    howyoudoing@centralperk.coffee  212-555-1005            N  
..                              ...           ...          ...  
516         agentp@danvillekids.com  555-555-3206            N  
517        baljeet@danvillekids.com  555-555-3207            N  
518         buford@danvillekids.com  555-555-3208            N  
519        vanessa@danvillekids.com  555-555-3209            N  
520          linda@danvillekids.com  555-555-3210            N  

[521 rows x 7 columns]

In [590]:
#for i in df1_cleaned["phone"]:
    print(i)   

212-555-1001
212-555-1002
212-555-1003
212-555-1004
212-555-1005
212-555-1006
212-555-1007
212-555-1008
212-555-1009
212-555-1010
505-555-2001
505-555-2002
505-555-2003
505-555-2004
505-555-2005
505-555-2006
505-555-2007
505-555-2008
505-555-2009
505-555-2010
123-555-3001
123-555-3002
123-555-3003
123-555-3004
123-555-3005
123-555-3006
123-555-3007
123-555-3008
123-555-3009
123-555-3010
636-555-4001
636-555-4002
636-555-4003
636-555-4004
636-555-4005
636-555-4006
636-555-4007
636-555-4008
636-555-4009
636-555-4010
570-555-5001
570-555-5002
570-555-5003
570-555-5004
570-555-5005
570-555-5006
570-555-5007
570-555-5008
570-555-5009
570-555-5010
219-555-6001
219-555-6002
219-555-6003
219-555-6004
219-555-6005
317-555-7001
317-555-7002
317-555-7003
317-555-7004
317-555-7005
718-555-8001
718-555-8002
718-555-8003
718-555-8004
718-555-8005
303-555-9001
303-555-9002
303-555-9003
303-555-9004
303-555-9005
303-555-9006
303-555-9007
303-555-9008
303-555-9009
303-555-9010
202-555-9101
202-555-9102

In [591]:
#print(len(i))

12


In [592]:
#for i in df1_cleaned["phone"]:
    if len(i) > 12:
        print(i)

1701-555-2201
1701-555-2202
1701-555-2203
1701-555-2204
1701-555-2205
1701-555-2206
1701-555-2207
1701-555-2208
1701-555-2209
1701-555-2210


In [600]:
#for i in df1_cleaned["phone"]:
    if len(i) > 12:
        print(i.lstrip("1"))
#dfl_cleaned = df1        

701-555-2201
701-555-2202
701-555-2203
701-555-2204
701-555-2205
701-555-2206
701-555-2207
701-555-2208
701-555-2209
701-555-2210


In [603]:
#df1_cleaned.head(289)

cust_id        date      time             name  \
0          1  2023-03-15  08:45:12           Rachel   
1          2  2023-05-22  12:30:45        R. Geller   
2          3  2023-07-09  18:15:27    Monica Geller   
3          4  2023-09-01  21:05:33    Chandler Bing   
4          5  2023-11-18  14:22:10             Joey   
..       ...         ...       ...              ...   
284      285  2024-03-15  12:10:30     Pirate Jenny   
285      286  2023-01-01  08:00:00  Jean-Luc Picard   
286      287  2023-02-14  12:25:15    William Riker   
287      288  2023-03-27  15:50:30             Data   
288      289  2023-05-10  19:15:45  Geordi La Forge   

                              email          phone sms-opt-out   
0         rachel@centralperk.coffee   212-555-1001            N  
1          rossg@centralperk.coffee   212-555-1002            N  
2     chefmonica@centralperk.coffee   212-555-1003            N  
3      chandlerb@centralperk.coffee   212-555-1004            Y  
4    howyoudoing@centralperk.coffee   212-555-1005            N  
..                              ...            ...          ...  
284        jenny@keenecounty.police   918-555-2110            N  
285     picard@enterprise.starfleet  1701-555-2201            N  
286      riker@enterprise.starfleet  1701-555-2202            N  
287       data@enterprise.starfleet  1701-555-2203            N  
288     geordi@enterprise.starfleet  1701-555-2204            N  

[289 rows x 7 columns]

In [548]:
#df1["phone"] = df1["phone"].str.replace("1701", "701", 9)
#df1

# I then replaced the characters in the string withogether since I seen that both of them are together for the first 50 records. 
# This time I replaced it with a period since I noticed they were in between a first intial followed by a last name
# I again pulled up the name column in brackets, this will specify the column I want to make changes to
# I then applied the .str method again to make sure the output we get in return is also a string, along with the. replace method
# I then went ahead and assigned it to the same dataframe and same column to save the changes
# Finally, I called on it to view the output

cust_id        date      time                  name  \
0          1  2023-03-15  08:45:12                Rachel   
1          2  2023-05-22  12:30:45             R. Geller   
2          3  2023-07-09  18:15:27         Monica Geller   
3          4  2023-09-01  21:05:33         Chandler Bing   
4          5  2023-11-18  14:22:10                  Joey   
..       ...         ...       ...                   ...   
516      517  2024-02-15  11:30:30    Perry the Platypus   
517      518  2024-05-01  15:55:45               Baljeet   
518      519  2023-03-05  09:20:00                Buford   
519      520  2023-05-18  13:45:15            Vanessa D.   
520      521  2023-08-01  18:10:30  Linda Flynn-Fletcher   

                              email         phone sms-opt-out   
0         rachel@centralperk.coffee  212-555-1001            N  
1          rossg@centralperk.coffee  212-555-1002            N  
2     chefmonica@centralperk.coffee  212-555-1003            N  
3      chandlerb@centralperk.coffee  212-555-1004            Y  
4    howyoudoing@centralperk.coffee  212-555-1005            N  
..                              ...           ...          ...  
516         agentp@danvillekids.com  555-555-3206            N  
517        baljeet@danvillekids.com  555-555-3207            N  
518         buford@danvillekids.com  555-555-3208            N  
519        vanessa@danvillekids.com  555-555-3209            N  
520          linda@danvillekids.com  555-555-3210            N  

[521 rows x 7 columns]

In [558]:
#df1.head(237)

cust_id        date      time           name  \
0          1  2023-03-15  08:45:12         Rachel   
1          2  2023-05-22  12:30:45      R. Geller   
2          3  2023-07-09  18:15:27  Monica Geller   
3          4  2023-09-01  21:05:33  Chandler Bing   
4          5  2023-11-18  14:22:10           Joey   
..       ...         ...       ...            ...   
232      233  2023-12-14  19:25:10   Hannah McKay   
233      234  2024-01-27  07:50:25   Harry Morgan   
234      235  2024-03-11  11:15:40   Lumen Pierce   
235      236  2023-03-03  08:45:15  Piper Chapman   
236      237  2023-04-16  13:10:30     Alex Vause   

                              email         phone sms-opt-out   
0         rachel@centralperk.coffee  212-555-1001            N  
1          rossg@centralperk.coffee  212-555-1002            N  
2     chefmonica@centralperk.coffee  212-555-1003            N  
3      chandlerb@centralperk.coffee  212-555-1004            Y  
4    howyoudoing@centralperk.coffee  212-555-1005            N  
..                              ...           ...          ...  
232           hannah@miamimetro.gov  305-555-1608            N  
233            harry@miamimetro.gov  305-555-1609            N  
234            lumen@miamimetro.gov  305-555-1610            N  
235            piper@litchfield.pen  607-555-1701            N  
236             alex@litchfield.pen  607-555-1702            N  

[237 rows x 7 columns]

In [557]:
#df1.head(363)

cust_id        date      time              name  \
0          1  2023-03-15  08:45:12            Rachel   
1          2  2023-05-22  12:30:45         R. Geller   
2          3  2023-07-09  18:15:27     Monica Geller   
3          4  2023-09-01  21:05:33     Chandler Bing   
4          5  2023-11-18  14:22:10              Joey   
..       ...         ...       ...               ...   
358      359  2024-06-11  19:35:25          Log Lady   
359      360  2023-02-17  08:00:50     Leland Palmer   
360      361  2023-05-02  13:25:15        Lucy Moran   
361      362  2023-02-10  08:15:25  Wednesday Addams   
362      363  2023-04-23  12:40:50   Morticia Addams   

                              email         phone sms-opt-out   
0         rachel@centralperk.coffee  212-555-1001            N  
1          rossg@centralperk.coffee  212-555-1002            N  
2     chefmonica@centralperk.coffee  212-555-1003            N  
3      chandlerb@centralperk.coffee  212-555-1004            Y  
4    howyoudoing@centralperk.coffee  212-555-1005            N  
..                              ...           ...          ...  
358            loglady@twinpeaks.wa  206-555-1608            N  
359             leland@twinpeaks.wa  206-555-1609            N  
360               lucy@twinpeaks.wa  206-555-1610            N  
361     wednesday@nevermore.academy  914-555-1701            N  
362      morticia@nevermore.academy  914-555-1702            N  

[363 rows x 7 columns]

In [619]:
df1_cleaned["phone"] = df_cleaned.phone.fillna(

NameError: name 'nan' is not defined